File: LLD/Python/railway/train.py

class Train:
    def __init__(self, number: str, name: str, total_seats: int):
        self.number = number
        self.name = name
        self.total_seats = total_seats
        self.booked = [False] * (total_seats + 1)  # 1-indexed

    def available_seats(self) -> int:
        return sum(1 for i in range(1, self.total_seats + 1) if not self.booked[i])

    def book_seat(self) -> int | None:
        for i in range(1, self.total_seats + 1):
            if not self.booked[i]:
                self.booked[i] = True
                return i
        return None

    def cancel_seat(self, seat_no: int) -> bool:
        if 1 <= seat_no <= self.total_seats and self.booked[seat_no]:
            self.booked[seat_no] = False
            return True
        return False

    def __repr__(self):
        return f"{self.number} - {self.name} (avail: {self.available_seats()})"

File: LLD/Java/railway/Train.java
public class Train {
    public final String number;
    public final String name;
    public final int totalSeats;
    private boolean[] booked;
    
    public Train(String number, String name, int totalSeats) {
        this.number = number; 
        this.name = name; 
        this.totalSeats = totalSeats;
        this.booked = new boolean[totalSeats + 1]; // 1-indexed
    }

    public int availableSeats() {
        int c = 0; 
        for (int i = 1; i <= totalSeats; i++) {
            if (!booked[i]) {
                c++;
            }
        }
        return c;
    }

    public int bookSeat() {
        for (int i = 1; i <= totalSeats; i++) {
            if (!booked[i]) {
                booked[i] = true;
                return i;
            }
        }
        return -1;
    }

    public boolean cancelSeat(int seatNo) {
        if (seatNo < 1 || seatNo > totalSeats || !booked[seatNo]) return false;
        booked[seatNo] = false; 
        return true;
    }
    
    @Override 
    public String toString() { 
        return number + " - " + name + " (avail: " + availableSeats() + ")"; 
    }
}
